### PR TITLE
Add container_mode to set source/service for container logs read from journald

### PR DESCRIPTION
**Pull request by vboulineau**
Add container_mode to set source/service for container logs read from journald

It takes the value from tags and use a cache as it's called for every log line fetched from journald.
We randomize the TTL in cache to avoid stalling the agent for some time in case of huge number of containers/tags.

### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [inline pkg/logs/input/journald/tailer_util.go]: why do we want to add an extra expiration time ? worth adding a comment to explain please

Comment (ajacquemot) [reply pkg/logs/input/journald/tailer_util.go]: if tags was a map, we would not need this cache I think, WDYT ? Also, do we really want to cache this ? what's the cardinality of tags ? Is it gonna be a huge gain ?

**Review by ajacquemot — COMMENTED** [review]
Thanks for submitting the PR, lot of new cool stuff that I did know about like that we had an internal `cache`. Left a bunch of comments, feel free to discard them if I am off topic.

File: pkg/logs/input/journald/tailer.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer.go]: Currently it always return "docker", and it was requested to add a parameter to enable new behaviour to keep backward compatibility.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Agree, I asked the same, apparently it's already aggregated/serialized when we get there, so apparently no easy way to get a map :/.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: I need to check that, though it's used like that for other types we put in cache. This code is the only one doing read/write on these cache keys, so I guess it should be pretty ok but I'll check with the team.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Probably indeed, tagger should ensure we only have 1 occurence of a tag key, not here.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Correct, if tag is map, then nothing to do. Earlier comment answers the why. Expiration is just to make sure we purge at some point when a docker is stopped/deleted. By default cache has a 5mins retention I believe, which is a bit low in this case as the docker_id <> image_name is never going to change.

Indeed, I asked and cardinality could be in ~100s in worst cases, which I benchmarked to be around ~300ns on my laptop. As the code is called for every single entry in journald, it could be causing high cpu usage on the agent if we have a heavy logger, so better avoid it.

I'll add a comment to explain this.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: IMHO, it's more a specific journald logic (that does not provide image information in fields we receive) than docker itself. It's not meant to be used elsewhere so I put it next to client code. But I agree we could see both ways, I don't mind moving it if there is a strong preference for that.

File: pkg/logs/input/journald/tailer_test.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_test.go]: Good idea, will do.

File: pkg/logs/input/journald/tailer.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer.go]: I think we don't need this config parameter and we always want to use the short image name when we can, so the code would look something like:
```
if t.isContainerEntry(entry) {
	if shortName, found := getDockerImageShortName(t.getContainerID(entry), tags); found {
		return shortName
	}

	return defaultApplicationName
}
```

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer_util.go]: Up to you, in general I am not a huge fan of `utils` or `commons` because the logic is very often tied to something specific which I think should be the name of file/package but reusable in different places but this is very subjective.

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer_util.go]: We could write a simple test case to confirm injecting a int to the cache, call the function and check what is returned.

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer_util.go]: No prob, resolving this conversation though.

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer_util.go]: Sounds good thanks, resolving the conversation.

**Review by ajacquemot — COMMENTED** [review]
Not convinced we need a configuration parameter, otherwise it's clean and nicely done 👍.

File: pkg/logs/input/journald/tailer.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer.go]: @NBParis What is your opinion about this one?

Comment (NBParis) [reply pkg/logs/input/journald/tailer.go]: We actually do because of backward compatibility.

If we were to change this it might be a breaking change for customers that have already pipelines defined on the existing source and service value.

So let's keep the parameter but let's add it everywhere in the documentation so that when people copy paste the examples they have it.

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Forgot to update it, done now

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Checked and it panics :/. So I changed the code to check the "cast" and a utest to validate it ignores the value from cache.

File: pkg/logs/input/journald/tailer.go
Comment (ajacquemot) [reply pkg/logs/input/journald/tailer.go]: sounds good 👍, resolving this conversation

File: pkg/logs/input/journald/tailer_test.go
Comment (ajacquemot) [inline pkg/logs/input/journald/tailer_test.go]: even though it won't change the result, we probably want to use the correct tag here, e.g. `short_image:testImage `

File: pkg/logs/input/journald/tailer_util.go
Comment (ajacquemot) [inline pkg/logs/input/journald/tailer_util.go]: Do we want to override the entry if the cast is not ok ? I think it's fine, worth adding a comment stating that the key is specific enough to avoid collision.

**Review by ajacquemot — COMMENTED** [review]
Looks good, worth waiting for the approval from someone from your team though.

File: pkg/logs/input/journald/tailer_test.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_test.go]: Changed

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: Added a comment to clearly state we'll override anyway

File: pkg/logs/input/journald/tailer.go
Comment (clamoriniere) [inline pkg/logs/input/journald/tailer.go]: Just a small nit: 
```suggestion
			if shortName, found := getDockerImageShortName(t.getContainerID(entry), tags); found {
```

File: pkg/logs/input/journald/tailer_util.go
Comment (clamoriniere) [inline pkg/logs/input/journald/tailer_util.go]: Just a small nit:
```suggestion
	if cacheContent, hit := cache.Cache.Get(cacheKey); hit {
```

File: pkg/logs/input/journald/tailer_util.go
Comment (clamoriniere) [inline pkg/logs/input/journald/tailer_util.go]: I think this line can be removed, if line 60 is `return shortName, true`

**Review by clamoriniere — COMMENTED** [review]
small nits, overall looks very good ! 👍

File: pkg/logs/input/journald/tailer_util.go
Comment (vboulineau) [reply pkg/logs/input/journald/tailer_util.go]: We could remove it and with line 60 being `	return shortName, shortName != ""`, but in my opinion it's not as clear
`

Comment (clamoriniere) [reply pkg/logs/input/journald/tailer_util.go]: Sorry, I didn't saw that in one case it returns `true` and the other `false`.
I think you can ignore my previous comment

**Review by clamoriniere — APPROVED** [review]

**Review by ajacquemot — APPROVED** [review]

**Comment by NBParis** [comment]
Can we make sure that we update the documentation for journald as well? :)

Basically we should recommend to have this flag set every time journald is used

**Pull request merged**
